internal/recon/md5.go

```diff
@@ -0,0 +1,165 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package recon
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/logg"
+
+	"github.com/sapcc/swift-health-exporter/internal/promhelper"
+)
+
+// md5Task implements the collector.collectorTask interface.
+type md5Task struct {
+	pathToReconExecutable string
+	hostTimeout           int
+	ctxTimeout            time.Duration
+
+	all        *promhelper.TypedDesc
+	errors     *promhelper.TypedDesc
+	matched    *promhelper.TypedDesc
+	notMatched *promhelper.TypedDesc
+}
+
+func newMD5Task(pathToReconExecutable string, hostTimeout int, ctxTimeout time.Duration) task {
+	return &md5Task{
+		hostTimeout:           hostTimeout,
+		ctxTimeout:            ctxTimeout,
+		pathToReconExecutable: pathToReconExecutable,
+		all: promhelper.NewGaugeTypedDesc(
+			"swift_cluster_md5_all",
+			"Sum of matched-, not matched hosts, and errors encountered while check hosts for md5sum(s) as reported by the swift-recon tool.", []string{"kind"}),
+		errors: promhelper.NewGaugeTypedDesc(
+			"swift_cluster_md5_errors",
+			"Errors encountered while checking hosts for md5sum(s) as reported by the swift-recon tool.", []string{"storage_ip", "kind"}),
+		matched: promhelper.NewGaugeTypedDesc(
+			"swift_cluster_md5_matched",
+			"Matched hosts for md5sum(s) reported by the swift-recon tool.", []string{"storage_ip", "kind"}),
+		notMatched: promhelper.NewGaugeTypedDesc(
+			"swift_cluster_md5_not_matched",
+			"Not matched hosts for md5sum(s) reported by the swift-recon tool.", []string{"storage_ip", "kind"}),
+	}
+}
+
+// describeMetrics implements the task interface.
+func (t *md5Task) describeMetrics(ch chan<- *prometheus.Desc) {
+	t.all.Describe(ch)
+	t.errors.Describe(ch)
+	t.matched.Describe(ch)
+	t.notMatched.Describe(ch)
+}
+
+// md5OutputBlockRx extracts the entire output block for a specific kind from
+// the aggregate md5 recon output.
+//
+// Match group ref:
+//  <1: kind> <2: output block>
+//
+// Example of an output block for ring:
+//   [<time-stamp>] Checking ring md5sums
+//   -> On disk object.ring.gz md5sum: 123456
+//   -> http://10.0.0.1:6000/recon/ringmd5: <urlopen error [Errno 111] ECONNREFUSED>
+//   -> http://10.0.0.2:6000/recon/ringmd5: {'/path/to/account.ring.gz': '123456', '/path/to/container.ring.gz': '123456', '/path/to/object.ring.gz': '123456'}
+//   -> http://10.0.0.2:6000/recon/ringmd5 matches.
+// 1/2 hosts matched, 1 error[s] while checking hosts.
+var md5OutputBlockRx = regexp.MustCompile(
+	`(?m)^.* Checking ([\.a-zA-Z0-9_]+) md5sums?\s*((?:(?:->|!!)\s\S*\s.*\s*)*)\s*[0-9]+/[0-9]+ hosts matched, [0-9]+ error.*$`)
```

Comment:
The middle part is really hard to decipher. Since we don't need most of the internal structure at this step, this could use some simplification:

```suggestion
	`(?m)^.* Checking ([\.a-zA-Z0-9_]+) md5sums?\s*((?:(?:->|!!).*\n)*)\s*[0-9]+/[0-9]+ hosts matched, [0-9]+ error.*$`)
```